static/js/composebox_typeahead.js

```diff
@@ -775,108 +806,140 @@ exports.content_typeahead_selected = function (item, event) {
-        const is_silent = this.completing === "silent_mention";
-        beginning = beginning.slice(0, -this.token.length - 1);
-        if (beginning.endsWith("@_*")) {
-            beginning = beginning.slice(0, -3);
-        } else if (beginning.endsWith("@*") || beginning.endsWith("@_")) {
-            beginning = beginning.slice(0, -2);
-        } else if (beginning.endsWith("@")) {
-            beginning = beginning.slice(0, -1);
-        }
-        if (user_groups.is_user_group(item)) {
-            beginning += "@*" + item.name + "* ";
-            // We could theoretically warn folks if they are
-            // mentioning a user group that literally has zero
-            // members where we are posting to, but we don't have
-            // that functionality yet, and we haven't gotten much
-            // feedback on this being an actual pitfall.
-        } else {
-            const mention_text = people.get_mention_syntax(item.full_name, item.user_id, is_silent);
-            beginning += mention_text + " ";
-            if (!is_silent) {
+    switch (this.completing) {
+        case "emoji":
+            // leading and trailing spaces are required for emoji,
+            // except if it begins a message or a new line.
+            if (
+                beginning.lastIndexOf(":") === 0 ||
+                beginning.charAt(beginning.lastIndexOf(":") - 1) === " " ||
+                beginning.charAt(beginning.lastIndexOf(":") - 1) === "\n"
+            ) {
+                beginning =
+                    beginning.slice(0, -this.token.length - 1) + ":" + item.emoji_name + ": ";
+            } else {
+                beginning =
+                    beginning.slice(0, -this.token.length - 1) + " :" + item.emoji_name + ": ";
+            }
+            break;
+        case "silent_mention":
+            beginning = beginning.slice(0, -this.token.length - 1);
+            if (beginning.endsWith("@_*")) {
+                beginning = beginning.slice(0, -3);
+            } else if (beginning.endsWith("@*") || beginning.endsWith("@_")) {
+                beginning = beginning.slice(0, -2);
+            } else if (beginning.endsWith("@")) {
+                beginning = beginning.slice(0, -1);
+            }
+            if (user_groups.is_user_group(item)) {
+                beginning += "@*" + item.name + "* ";
+                // We could theoretically warn folks if they are
+                // mentioning a user group that literally has zero
+                // members where we are posting to, but we don't have
+                // that functionality yet, and we haven't gotten much
+                // feedback on this being an actual pitfall.
+            } else {
+                const mention_text = people.get_mention_syntax(item.full_name, item.user_id, true); // is_silent = true
+                beginning += mention_text + " ";
+            }
+            break;
+        case "mention":
```

Comment:
You don't need to duplicate this block; just put `case "mention"` on the line after the `silent_mention` case, and it'll use the same implementation for both cases.